src/Features/Core/Portable/CodeRefactorings/ConvertNumericLiteral/AbstractConvertNumericLiteralCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,166 @@
+            }
+
+            if (kind != NumericKind.Binary)
+            {
+                RegisterRefactoringWithResult(binaryPrefix + Convert.ToString(value, 2), FeaturesResources.Convert_to_binary);
+            }
+
+            if (kind != NumericKind.Hexadecimal)
+            {
+                RegisterRefactoringWithResult(hexPrefix + value.ToString("X"), FeaturesResources.Convert_to_hex);
+            }
+
+            const string DigitSeparator = "_";
+            if (numericText.Contains(DigitSeparator))
+            {
+                RegisterRefactoringWithResult(prefix + number.Replace(DigitSeparator, string.Empty), FeaturesResources.Remove_separators);
+            }
+            else
+            {
+                switch (kind)
+                {
+                    case NumericKind.Decimal:
+                        RegisterRefactoringWithResult(AddSeparators(number, 3), FeaturesResources.Separate_thousands);
+                        break;
+
+                    case NumericKind.Hexadecimal:
+                        RegisterRefactoringWithResult(hexPrefix + AddSeparators(number, 4), FeaturesResources.Separate_words);
+                        break;
+
+                    case NumericKind.Binary:
+                        RegisterRefactoringWithResult(binaryPrefix + AddSeparators(number, 4), FeaturesResources.Separate_nibbles);
+                        break;
+                }
+            }
+
+            void RegisterRefactoringWithResult(string text, string title)
+            {
+                var generator = SyntaxGenerator.GetGenerator(document);
+                var updatedToken = generator.NumericLiteralToken(text + suffix, (ulong)value);
+                var updatedRoot = root.ReplaceToken(numericToken, updatedToken);
+                context.RegisterRefactoring(new MyCodeAction(title, c => Task.FromResult(document.WithSyntaxRoot(updatedRoot))));
+            }
+        }
+
+        private static (string prefix, string number, string suffix) GetNumericLiteralParts(string numericText, string hexPrefix, string binaryPrefix)
+        {
+            // Match literal text and extract out base prefix, type suffix and the number itself.
+            var groups = Regex.Match(numericText, $"({hexPrefix}|{binaryPrefix})?([_0-9a-f]+)(.*)", RegexOptions.IgnoreCase).Groups;
+            return (prefix: groups[1].Value, number: groups[2].Value, suffix: groups[3].Value);
+        }
+
+        private static string AddSeparators(string numericText, int interval)
+        {
+            // Insert digit separators in the given interval.
+            var result = Regex.Replace(numericText, $"(.{{{interval}}})", "_$1", RegexOptions.RightToLeft);
+            // Fix for the case "0x_1111" that is not supported yet.
+            return result[0] == '_' ? result.Substring(1) : result;
+        }
+
+        private static bool IsIntegral(SpecialType specialType)
```

Comment:
`Microsoft.CodeAnalysis.SpecialTypeExtensions` is not accessible, however, `Microsoft.CodeAnalysis.CSharp.Symbols.SpecialTypeExtensions.IsIntegralType` is.